### PR TITLE
WIP: Fastify v3 upgrade

### PR DESCRIPTION
**Pull request by thomheymann**
### TODO

- [] bump fastify version in index.js
- [] use published fastify/fastify-plugin/middie dependencies once published

#### Checklist

- [ ] run `npm run test` and `npm run benchmark`
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -1,197 +1,215 @@
...
 
-  interface IHelmetHidePoweredByConfiguration {
-    setTo?: string;
-  }
+export interface IHelmetHidePoweredByConfiguration {
+  setTo?: string;
+}
 
-  interface IHelmetSetIfFunction {
-    (req: fastify.FastifyRequest<http.IncomingMessage>, res: fastify.FastifyReply<http.ServerResponse>): boolean;
```
Comment (thomheymann) [inline index.d.ts:164]: This type assumed http request which is not always true

File: index.d.ts [hunk 2]
```diff
@@ -1,197 +1,215 @@
...
+  ieNoOpen?: boolean;
+  noCache?: boolean;
+  noSniff?: boolean;
+  permittedCrossDomainPolicies?: boolean | IHelmetPermittedCrossDomainPoliciesConfiguration;
+  referrerPolicy?: boolean | IHelmetReferrerPolicyConfiguration;
+  xssFilter?: boolean | IHelmetXssFilterConfiguration;
 }
+
+// We dont know the server type so keep using fastify.RawServerBase (union of all server types) for FastifyHelmetOptions
+declare const fastifyHelmet: FastifyPlugin<FastifyHelmetOptions<fastify.RawServerBase>>;
```
Comment (thomheymann) [inline index.d.ts:214]: We dont know the server type at the time of authoring the plugin so using `fastify.RawServerBase` (which is a union of all server types) for `FastifyHelmetOptions`.

Users of the plugin should then be able to cast req / res to a specific http type

File: index.test-d.ts [hunk 3]
```diff
@@ -1,24 +1,35 @@
...
+
+const app2 = fastify({ http2: true });
+app2.register(fastifyHelmet, {});
+app2.register(fastifyHelmet, {
+  frameguard: {
+    action: "deny"
+  },
+  contentSecurityPolicy: {
+    directives: {
+      scriptSrc: ["scripts.example.com", (req: FastifyRequest<http2.Http2Server>, res: FastifyReply<http2.Http2Server>) => {
```
Comment (thomheymann) [inline index.test-d.ts:22]: Idea was to let users cast to server specific request object but that fails:

```
  index.test-d.ts:16:29
  ✖   16:29  Argument of type { frameguard: { action: string; }; contentSecurityPolicy: { directives: { scriptSrc: (string | ((req: FastifyRequest<http2.Http2Server, http2.Http2ServerRequest, RequestGenericInterface>, res: FastifyReply<...>) => string))[]; }; }; } is not assignable to parameter of type (RegisterOptions & FastifyHelmetOptions<RawServerBase, IncomingMessage | Http2ServerRequest, ServerResponse | Http2ServerResponse>) | (() => RegisterOptions & FastifyHelmetOptions<...>) | undefined.
  Type { frameguard: { action: string; }; contentSecurityPolicy: { directives: { scriptSrc: (string | ((req: FastifyRequest<http2.Http2Server, http2.Http2ServerRequest, RequestGenericInterface>, res: FastifyReply<...>) => string))[]; }; }; } is not assignable to type RegisterOptions & FastifyHelmetOptions<RawServerBase, IncomingMessage | Http2ServerRequest, ServerResponse | Http2ServerResponse>.
    Type { frameguard: { action: string; }; contentSecurityPolicy: { directives: { scriptSrc: (string | ((req: FastifyRequest<http2.Http2Server, http2.Http2ServerRequest, RequestGenericInterface>, res: FastifyReply<...>) => string))[]; }; }; } is not assignable to type FastifyHelmetOptions<RawServerBase, IncomingMessage | Http2ServerRequest, ServerResponse | Http2ServerResponse>.
      Types of property contentSecurityPolicy are incompatible.
        Type { directives: { scriptSrc: (string | ((req: FastifyRequest<http2.Http2Server, http2.Http2ServerRequest, RequestGenericInterface>, res: FastifyReply<...>) => string))[]; }; } is not assignable to type boolean | IHelmetContentSecurityPolicyConfiguration<RawServerBase, IncomingMessage | Http2ServerRequest, ServerResponse | Http2ServerResponse> | undefined.
          Type { directives: { scriptSrc: (string | ((req: FastifyRequest<http2.Http2Server, http2.Http2ServerRequest, RequestGenericInterface>, res: FastifyReply<...>) => string))[]; }; } is not assignable to type IHelmetContentSecurityPolicyConfiguration<RawServerBase, IncomingMessage | Http2ServerRequest, ServerResponse | Http2ServerResponse>.
            The types of directives.scriptSrc are incompatible between these types.
              Type (string | ((req: FastifyRequest<Http2Server, Http2ServerRequest, RequestGenericInterface>, res: FastifyReply<...>) => string))[] is not assignable to type HelmetCspDirectiveValue<RawServerBase, IncomingMessage | Http2ServerRequest, ServerResponse | Http2ServerResponse>[].
                Type string | ((req: FastifyRequest<Http2Server, Http2ServerRequest, RequestGenericInterface>, res: FastifyReply<...>) => string) is not assignable to type HelmetCspDirectiveValue<RawServerBase, IncomingMessage | Http2ServerRequest, ServerResponse | Http2ServerResponse>.
                  Type (req: FastifyRequest<http2.Http2Server, http2.Http2ServerRequest, RequestGenericInterface>, res: FastifyReply<http2.Http2Server, http2.Http2ServerResponse, unknown>) => string is not assignable to type HelmetCspDirectiveValue<RawServerBase, IncomingMessage | Http2ServerRequest, ServerResponse | Http2ServerResponse>.
                    Type (req: FastifyRequest<http2.Http2Server, http2.Http2ServerRequest, RequestGenericInterface>, res: FastifyReply<http2.Http2Server, http2.Http2ServerResponse, unknown>) => string is not assignable to type IHelmetContentSecurityPolicyDirectiveFunction<RawServerBase, IncomingMessage | Http2ServerRequest, ServerResponse | Http2ServerResponse>.
                      Types of parameters req and req are incompatible.
                        Type FastifyRequest<RawServerBase, IncomingMessage | Http2ServerRequest, RequestGenericInterface> is not assignable to type FastifyRequest<Http2Server, Http2ServerRequest, RequestGenericInterface>.
                          Type IncomingMessage & FastifyRequestInterface<RawServerBase, IncomingMessage | Http2ServerRequest, RequestGenericInterface> is not assignable to type FastifyRequest<Http2Server, Http2ServerRequest, RequestGenericInterface>.
                            Type IncomingMessage & FastifyRequestInterface<RawServerBase, IncomingMessage | Http2ServerRequest, RequestGenericInterface> is missing the following properties from type Http2ServerRequest: authority, scheme, stream  
```

File: index.test-d.ts [hunk 4]
```diff
@@ -97,19 +108,19 @@ function contentSecurityPolicyTest() {
...
+    _req: FastifyRequest<http.Server>,
+    _res: FastifyReply<http.Server>
   ) {
     return false;
   }
 
-  app.register(fastifyHelmet, {
+  const consigWithCallbacks: FastifyHelmetOptions<http.Server> = {
     contentSecurityPolicy: {
       directives: {
```
Comment (thomheymann) [inline index.test-d.ts:125]: Fails with:

```
  ✖  138:30  Argument of type FastifyHelmetOptions<Server, IncomingMessage, ServerResponse> is not assignable to parameter of type (RegisterOptions & FastifyHelmetOptions<RawServerBase, IncomingMessage | Http2ServerRequest, ServerResponse | Http2ServerResponse>) | (() => RegisterOptions & FastifyHelmetOptions<...>) | undefined.
  Type FastifyHelmetOptions<Server, IncomingMessage, ServerResponse> is not assignable to type RegisterOptions & FastifyHelmetOptions<RawServerBase, IncomingMessage | Http2ServerRequest, ServerResponse | Http2ServerResponse>.
    Type FastifyHelmetOptions<Server, IncomingMessage, ServerResponse> is not assignable to type FastifyHelmetOptions<RawServerBase, IncomingMessage | Http2ServerRequest, ServerResponse | Http2ServerResponse>.
      Type RawServerBase is not assignable to type Server.
        Type Http2Server is missing the following properties from type Server: maxHeadersCount, timeout, headersTimeout, keepAliveTimeout    
```

File: index.js [hunk 5]
```diff
@@ -22,25 +22,28 @@ const helmet = {
...
+            fastify.use(middleware(option))
+          }
+        } else if (isDefault) {
+          fastify.use(middleware({}))
+        }
+      })
+      next()
+    })
 }, {
   fastify: '>=0.39.1',
```
Comment (mcollina) [inline index.js:48]: This needs to be bumped.

Comment (thomheymann) [reply index.js]: It would fail the build and i want the tests passing
I'll add it as a todo

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -0,0 +1,18 @@
+name: CI workflow
+on: [push, pull_request]
+jobs:
+  test:
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        node-version: [10.x, 12.x, 13.x]
```
Comment (mcollina) [inline .github/workflows/ci.yml:8]:
```suggestion
        node-version: [10.x, 12.x, 14.x]
```

Comment (mcollina) [reply .github/workflows/ci.yml:8]: this is still not done

File: index.js [hunk 7]
```diff
@@ -22,26 +22,38 @@ const helmet = {
 const middlewares = Object.keys(helmet)
 
 module.exports = fp(function (fastify, opts, next) {
-  middlewares.forEach(function (middlewareName) {
-    const middleware = helmet[middlewareName]
-    const option = opts[middlewareName]
-    const isDefault = config.defaultMiddleware.indexOf(middlewareName) !== -1
+  function useMiddlewares () {
+    middlewares.forEach(function (middlewareName) {
```
Comment (mcollina) [inline index.js:26]: Please use a `for..of` loop instead.

**Comment by Ethan-Arrowood** [comment]
@thomheymann do you have publishing permissions? If not I can merge and publish

**Comment by mcollina** [comment]
go ahead and publish Ethan, Thom is not a member of this org.

**Pull request merged**